File: OnlyCueTests/LTCScheduleGoldenVectorTests.swift
import XCTest
@testable import OnlyCue

// The generator + drift guard for `golden/ltc-schedule-v1.json`. The contract
// model and the case matrix live in `LTCScheduleGolden.swift`.

extension LTCScheduleGolden {

    // MARK: - Case builders

    private static func streamCase(_ spec: StreamCase) throws -> LTCScheduleGoldenVector.Case {
        let stream = try stream(spec.rate, spec.sampleRate, start: streamTimecode)
        let samples = stream.samples(frameCount: spec.frameCount)
        let joins = stride(from: stream.samplesPerFrame, to: samples.count, by: stream.samplesPerFrame).map {
            LTCScheduleGoldenVector.Join(at: $0, runs: runs(of: window(of: samples, around: $0)))
        }
        return LTCScheduleGoldenVector.Case(
            label: "stream/\(spec.rate.rawValue)@\(Int(spec.sampleRate))/\(stream.startTimecode.displayString)"
                + "/x\(spec.frameCount)",
            op: "stream",
            rate: spec.rate.rawValue,
            input: LTCScheduleGoldenVector.Input(
                hours: stream.startTimecode.hours,
                minutes: stream.startTimecode.minutes,
                seconds: stream.startTimecode.seconds,
                frames: stream.startTimecode.frames,
                sampleRate: GoldenDouble(spec.sampleRate),
                amplitude: GoldenDouble(Double(amplitude)),
                frameCount: spec.frameCount
            ),
            expect: LTCScheduleGoldenVector.Expect(
                samplesPerFrame: stream.samplesPerFrame,
                totalSamples: samples.count,
                firstSampleIsHigh: samples.first.map { $0 > 0 },
                lastSampleIsHigh: samples.last.map { $0 > 0 },
                joins: joins
            )
        )
    }

    private static func streamTimecodeCase(_ rate: SMPTEFramerate, offset: Int) throws -> LTCScheduleGoldenVector.Case {
        let stream = try stream(rate, 48000, start: frameOffsetOrigin)
        return LTCScheduleGoldenVector.Case(
            label: "streamTimecode/\(rate.rawValue)/\(stream.startTimecode.displayString)/offset\(offset)",
            op: "streamTimecode",
            rate: rate.rawValue,
            input: LTCScheduleGoldenVector.Input(
                hours: stream.startTimecode.hours,
                minutes: stream.startTimecode.minutes,
                seconds: stream.startTimecode.seconds,
                frames: stream.startTimecode.frames,
                frameOffset: offset
            ),
            expect: LTCScheduleGoldenVector.Expect(
                timecode: stream.timecode(atFrameOffset: offset).displayString
            )
        )
    }

    private static func input(_ spec: ScheduleCase, timecode: Timecode) -> LTCScheduleGoldenVector.Input {
        LTCScheduleGoldenVector.Input(
            hours: timecode.hours,
            minutes: timecode.minutes,
            seconds: timecode.seconds,
            frames: timecode.frames,
            sampleRate: GoldenDouble(spec.sampleRate),
            amplitude: GoldenDouble(Double(amplitude)),
            framesPerBuffer: spec.framesPerBuffer
        )
    }

    private static func label(_ spec: ScheduleCase, timecode: Timecode) -> String {
        "\(spec.rate.rawValue)@\(Int(spec.sampleRate))/\(timecode.displayString)/fpb\(spec.framesPerBuffer)"
    }

    private static func scheduleCase(_ spec: ScheduleCase) throws -> LTCScheduleGoldenVector.Case {
        let schedule = try schedule(spec)
        return LTCScheduleGoldenVector.Case(
            label: "schedule/" + label(spec, timecode: schedule.startTimecode),
            op: "schedule",
            rate: spec.rate.rawValue,
            input: input(spec, timecode: schedule.startTimecode),
            expect: LTCScheduleGoldenVector.Expect(
                samplesPerBuffer: schedule.samplesPerBuffer,
                bufferDuration: GoldenDouble(schedule.bufferDuration)
            )
        )
    }

    private static func bufferCase(_ spec: BufferCase) throws -> LTCScheduleGoldenVector.Case {
        let schedule = try schedule(spec.schedule)
        let buffer = schedule.buffer(at: spec.index)
        var input = input(spec.schedule, timecode: schedule.startTimecode)
        input.bufferIndex = spec.index
        return LTCScheduleGoldenVector.Case(
            label: "buffer/" + label(spec.schedule, timecode: schedule.startTimecode) + "/index\(spec.index)",
            op: "buffer",
            rate: spec.schedule.rate.rawValue,
            input: input,
            expect: LTCScheduleGoldenVector.Expect(
                lastSampleIsHigh: buffer.samples.last.map { $0 > 0 },
                timecode: buffer.timecode.displayString,
                sampleCount: buffer.samples.count
            )
        )
    }

    /// The run lengths of `last(joinWindow) ++ first(joinWindow)` across the join
    /// between buffer `index` and `index + 1`. A port whose buffers do not abut
    /// cleanly produces a doubled run in the middle of this array.
    private static func bufferSeamCase(_ spec: BufferCase) throws -> LTCScheduleGoldenVector.Case {
        let schedule = try schedule(spec.schedule)
        let previous = schedule.samples(forBufferIndex: spec.index)
        let next = schedule.samples(forBufferIndex: spec.index + 1)
        let seam = previous.suffix(joinWindow) + next.prefix(joinWindow)
        var input = input(spec.schedule, timecode: schedule.startTimecode)
        input.bufferIndex = spec.index
        return LTCScheduleGoldenVector.Case(
            label: "bufferSeam/" + label(spec.schedule, timecode: schedule.startTimecode)
                + "/index\(spec.index)-\(spec.index + 1)",
            op: "bufferSeam",
            rate: spec.schedule.rate.rawValue,
            input: input,
            expect: LTCScheduleGoldenVector.Expect(
                previousEndsHigh: previous.last.map { $0 > 0 },
                nextStartsHigh: next.first.map { $0 > 0 },
                runs: runs(of: Array(seam))
            )
        )
    }

    private static func targetCountCase(_ spec: TargetCountCase) throws -> LTCScheduleGoldenVector.Case {
        let schedule = try schedule(spec.schedule)
        var input = input(spec.schedule, timecode: schedule.startTimecode)
        input.elapsedSeconds = GoldenDouble(spec.elapsedSeconds)
        input.leadBuffers = spec.leadBuffers
        return LTCScheduleGoldenVector.Case(
            label: "targetBufferCount/" + label(spec.schedule, timecode: schedule.startTimecode)
                + "/elapsed\(spec.elapsedSeconds)/lead\(spec.leadBuffers)",
            op: "targetBufferCount",
            rate: spec.schedule.rate.rawValue,
            input: input,
            expect: LTCScheduleGoldenVector.Expect(
                count: schedule.targetBufferCount(
                    elapsedSeconds: spec.elapsedSeconds,
                    leadBuffers: spec.leadBuffers
                )
            )
        )
    }

    private static func framesPerBufferCase(
        _ rate: SMPTEFramerate,
        targetSeconds: Double
    ) -> LTCScheduleGoldenVector.Case {
        LTCScheduleGoldenVector.Case(
            label: "framesPerBuffer/\(rate.rawValue)/\(targetSeconds)s",
            op: "framesPerBuffer",
            rate: rate.rawValue,
            input: LTCScheduleGoldenVector.Input(targetSeconds: GoldenDouble(targetSeconds)),
            expect: LTCScheduleGoldenVector.Expect(
                frames: LTCSchedule.framesPerBuffer(forTargetSeconds: targetSeconds, rate: rate)
            )
        )
    }

    // MARK: - Assembly

    static func make() throws -> LTCScheduleGoldenVector {
        var cases = try streamCases.map(streamCase)
        cases += try allRates.flatMap { rate in
            try frameOffsets.map { try streamTimecodeCase(rate, offset: $0) }
        }
        cases += try scheduleCases.map(scheduleCase)
        cases += try bufferCases.map(bufferCase)
        cases += try bufferSeamCases.map(bufferSeamCase)
        cases += try targetCountCases.map(targetCountCase)
        cases += allRates.flatMap { rate in
            targetSecondsValues.map { framesPerBufferCase(rate, targetSeconds: $0) }
        }
        return LTCScheduleGoldenVector(
            contract: "ltc-schedule",
            version: 1,
            note: "macOS-generated golden vectors for the OnlyCue LTC scheduling layer "
                + "(LTCFrameStream and LTCSchedule). The C# OnlyCue.Core re-implementation "
                + "must reproduce every case exactly. PCM is pinned by structure plus a "
                + "±16-sample run-length window around each join; every sample of a single "
                + "frame is already pinned by golden/ltc-wire-v1.json. To regenerate after a "
                + "deliberate change, delete golden/ltc-schedule-v1.json and re-run the suite.",
            cases: cases
        )
    }

    /// Deterministic encoding so the committed file is stable and diff-friendly.
    static func encoded(_ vector: LTCScheduleGoldenVector) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(vector)
    }
}

// MARK: - Tests

final class LTCScheduleGoldenVectorTests: XCTestCase {

    private func goldenURL() throws -> URL {
        try repoRoot().appendingPathComponent("golden", isDirectory: true)
            .appendingPathComponent("ltc-schedule-v1.json")
    }

    /// **The invariant this whole slice exists for.**
    ///
    /// `LTCSchedule` builds each buffer from a *fresh* `LTCFrameStream`, so
    /// biphase polarity is explicitly not threaded across buffer joins. Its doc
    /// comment used to call that harmless "since an LTC reader keys on
    /// transitions, not absolute polarity" — wrong reasoning, because a reset
    /// that actually flipped the level would mean a *missing* transition, which
    /// is a misread bit rather than an inversion.
    ///
    /// The real reason it is harmless: each frame emits 80 bit-boundary flips
    /// plus one mid-bit flip per `1` bit, and the parity bit forces an even
    /// number of ones — so the flip count is even and `endLevel` always equals
    /// `startLevel`. **The parity bit is what makes the buffer joins seamless.**
    /// A port with wrong parity would glitch audibly at every boundary.
    func test_endLevelAlwaysEqualsStartLevel_becauseParityMakesTheFlipCountEven() throws {
        for spec in LTCScheduleGolden.streamCases where spec.frameCount > 0 {
            let timecode = try LTCScheduleGolden.timecode(LTCScheduleGolden.streamTimecode, rate: spec.rate)
            for startLevel in [false, true] {
                let (samples, endLevel) = LTCEncoder.samples(
                    for: timecode,
                    sampleRate: spec.sampleRate,
                    amplitude: LTCScheduleGolden.amplitude,
                    startLevel: startLevel
                )
                XCTAssertEqual(endLevel, startLevel, "\(spec.rate.rawValue)@\(spec.sampleRate)")
                XCTAssertEqual(samples.last.map { $0 > 0 }, endLevel, "the last slot is emitted at the final level")
                XCTAssertEqual(LTCFrame(timecode: timecode).hasEvenParity, true, "even parity is why")
            }
        }
    }

    /// Independent pins that do **not** go through the generator, so a wrong
    /// implementation fails here rather than silently baking a wrong "golden"
    /// value. Hand-computed, not read off the implementation.
    func test_knownScheduleValues() throws {
        // `samplesPerBuffer` rounds per frame and then multiplies. 44100 / 24 is
        // 1837.5 → 1838 → ×4 = 7352. Multiplying first would give
        // round(4 × 1837.5) = 7350, i.e. two samples short per buffer forever.
        let schedule = LTCSchedule(
            startTimecode: try LTCScheduleGolden.timecode([1, 0, 0, 0], rate: .fps24),
            sampleRate: 44100,
            framesPerBuffer: 4
        )
        XCTAssertEqual(schedule.samplesPerBuffer, 7352)
        XCTAssertNotEqual(schedule.samplesPerBuffer, Int((4 * 44100.0 / 24.0).rounded()))
        XCTAssertEqual(schedule.samples(forBufferIndex: 0).count, 7352)
        XCTAssertEqual(schedule.bufferDuration, 4.0 / 24.0)

        // 0.1 s at 25 fps is 2.5 frames. Swift's half-away-from-zero rounding
        // gives 3; C#'s banker's `Math.Round` gives 2. 0.3 s is 7.5, where both
        // modes give 8 — the control.
        XCTAssertEqual(LTCSchedule.framesPerBuffer(forTargetSeconds: 0.1, rate: .fps25), 3)
        XCTAssertEqual(LTCSchedule.framesPerBuffer(forTargetSeconds: 0.5, rate: .fps25), 13)
        XCTAssertEqual(LTCSchedule.framesPerBuffer(forTargetSeconds: 0.3, rate: .fps25), 8)
        XCTAssertEqual(LTCSchedule.framesPerBuffer(forTargetSeconds: -1, rate: .fps25), 1)

        // `ceil`, not `round`: a fifth of a buffer still needs a whole buffer.
        XCTAssertEqual(schedule.targetBufferCount(elapsedSeconds: 0.2 * schedule.bufferDuration, leadBuffers: 0), 1)
        XCTAssertEqual(schedule.targetBufferCount(elapsedSeconds: -1, leadBuffers: -1), 0)
    }

    /// The tie cases have to *be* ties, or the contract goes quiet without
    /// failing. Pinned separately so a future edit to the matrix cannot silently
    /// remove the only cases that separate the two rounding modes.
    func test_theRoundingTrapCasesAreActuallyTies() {
        XCTAssertEqual(0.1 * 25.0, 2.5)
        XCTAssertEqual(0.5 * 25.0, 12.5)
        XCTAssertEqual(44100.0 / 24.0, 1837.5)
        XCTAssertTrue(LTCScheduleGolden.targetSecondsValues.contains(0.1))
        XCTAssertTrue(LTCScheduleGolden.targetSecondsValues.contains(0.5))
        XCTAssertTrue(LTCScheduleGolden.scheduleCases.contains { $0.sampleRate == 44100 && $0.rate == .fps24 })
    }

    /// Two mutants at a frame join, and the contract's honest answer to each.
    ///
    /// **Resetting the polarity at every frame is an _equivalent_ mutant.** Every
    /// frame ends at the level it started at — see
    /// `test_endLevelAlwaysEqualsStartLevel_…` — so the threaded `level` is
    /// `false` at the top of every iteration anyway, and a port that dropped the
    /// threading entirely would emit byte-identical PCM. Nothing can catch it,
    /// because there is nothing to catch. Asserted rather than assumed, so no
    /// future comment can claim the `joins` windows guard the threading.
    ///
    /// **Inverting the polarity at a join is not equivalent**, and that is what
    /// the window is sized for: it loses a transition, which is a misread bit
    /// rather than a harmless inversion. The whole second frame flips sign, so
    /// the assertion is that the difference is visible *inside* the ±`joinWindow`
    /// slice — a window too narrow to straddle the join would pass while pinning
    /// nothing.
    func test_joinWindow_catchesAnInvertedJoin_butResettingOneIsEquivalent() throws {
        /// The stream's own loop, with the level at each frame chosen by `level`.
        func concatenated(
            _ stream: LTCFrameStream,
            frameCount: Int,
            level: (_ previousEndLevel: Bool) -> Bool
        ) -> [Float] {
            var output: [Float] = []
            var next = false
            for offset in 0..<frameCount {
                let (samples, endLevel) = LTCEncoder.samples(
                    for: stream.timecode(atFrameOffset: offset),
                    sampleRate: stream.sampleRate,
                    amplitude: stream.amplitude,
                    startLevel: next
                )
                output += samples
                next = level(endLevel)
            }
            return output
        }

        for spec in LTCScheduleGolden.streamCases where spec.frameCount > 1 {
            let stream = try LTCScheduleGolden.stream(
                spec.rate, spec.sampleRate, start: LTCScheduleGolden.streamTimecode
            )
            let threaded = stream.samples(frameCount: spec.frameCount)
            let label = "\(spec.rate.rawValue)@\(Int(spec.sampleRate))/x\(spec.frameCount)"

            XCTAssertEqual(
                concatenated(stream, frameCount: spec.frameCount) { _ in false },
                threaded,
                "\(label): resetting the polarity per frame is indistinguishable from threading it"
            )

            let inverted = concatenated(stream, frameCount: spec.frameCount) { !$0 }
            XCTAssertEqual(inverted.count, threaded.count, label)
            let at = stream.samplesPerFrame
            XCTAssertNotEqual(
                LTCScheduleGolden.runs(of: LTCScheduleGolden.window(of: threaded, around: at)),
                LTCScheduleGolden.runs(of: LTCScheduleGolden.window(of: inverted, around: at)),
                "\(label): an inverted join must be visible inside the pinned window"
            )
        }
    }

    /// Run-length encoding is lossless for a `±amplitude` signal, so the window
    /// runs are equivalent to the samples they stand for.
    func test_runLengthEncoding_roundTripsTheSamples() throws {
        let stream = try LTCScheduleGolden.stream(.fps24, 48000, start: LTCScheduleGolden.streamTimecode)
        let samples = stream.samples(frameCount: 2)
        let window = LTCScheduleGolden.window(of: samples, around: stream.samplesPerFrame)
        let expanded = LTCScheduleGolden.runs(of: window).flatMap { run in
            repeatElement(Float(run[0]) * LTCScheduleGolden.amplitude, count: run[1])
        }
        XCTAssertEqual(expanded, window)
        XCTAssertEqual(window.count, 2 * LTCScheduleGolden.joinWindow)
    }

    /// Drift guard + bootstrap. When the committed file is present it must equal
    /// the current Swift output byte-for-byte (that's the contract the C# core
    /// verifies against). When it is missing — first run, or after a deliberate
    /// change where the dev deleted it to regenerate — it is written and the test
    /// fails, so the new contract is committed and reviewed rather than silently
    /// accepted in CI.
    func test_committedVectors_matchCurrentSwiftOutput() throws {
        let url = try goldenURL()
        let generated = try LTCScheduleGolden.encoded(LTCScheduleGolden.make())

        guard FileManager.default.fileExists(atPath: url.path) else {
            try FileManager.default.createDirectory(
                at: url.deletingLastPathComponent(), withIntermediateDirectories: true
            )
            try generated.write(to: url)
            return XCTFail("golden/ltc-schedule-v1.json was missing — generated it from the Swift "
                + "LTCFrameStream / LTCSchedule implementation. Commit the file and re-run. (To "
                + "regenerate after a deliberate change, delete the file first.)")
        }

        let committed = try JSONDecoder().decode(
            LTCScheduleGoldenVector.self, from: Data(contentsOf: url)
        )
        XCTAssertEqual(
            committed,
            try LTCScheduleGolden.make(),
            "golden/ltc-schedule-v1.json drifted from the Swift implementation. If the change "
                + "was intentional, delete the file and re-run to regenerate, review the "
                + "diff, and update the C# OnlyCue.Core to match."
        )
    }
}

File: windows/OnlyCue.Core.Tests/LtcScheduleGoldenVectorTests.cs
using OnlyCue.Core.Ltc;
using Xunit;

namespace OnlyCue.Core.Tests;

/// <summary>
/// The Windows half of the LTC scheduling contract (epic #728, M3 slice 2).
/// macOS emits <c>golden/ltc-schedule-v1.json</c> from the Swift
/// <c>LTCFrameStream</c> / <c>LTCSchedule</c>; this suite asserts the C#
/// re-implementation reproduces every case exactly.
/// </summary>
/// <remarks>
/// Slice 1 pinned every sample of a single frame. This one pins composition, so
/// PCM appears only as run lengths in a ±16-sample window around each join —
/// wide enough that a port which restarted the biphase polarity at every frame
/// boundary produces a visibly different window.
/// </remarks>
public class LtcScheduleGoldenVectorTests
{
    private static readonly LtcScheduleVector Vector = LtcScheduleVector.Load();

    /// <summary>Must match <c>LTCScheduleGolden.joinWindow</c> on the Swift side;
    /// <see cref="SeamWindow_IsTheWidthTheContractWasGeneratedAt"/> proves it.</summary>
    private const int JoinWindow = 16;

    public static TheoryData<string> CaseLabels
    {
        get
        {
            var data = new TheoryData<string>();
            foreach (var label in Vector.Cases.Select(c => c.Label))
            {
                data.Add(label);
            }

            return data;
        }
    }

    [Fact]
    public void GoldenFile_IsTheExpectedContract()
    {
        Assert.Equal("ltc-schedule", Vector.Contract);
        Assert.Equal(1, Vector.Version);
        Assert.NotEmpty(Vector.Cases);
    }

    [Theory]
    [MemberData(nameof(CaseLabels))]
    public void CSharpCore_ReproducesGoldenCase(string label)
    {
        var goldenCase = Vector.Cases.Single(c => c.Label == label);

        switch (goldenCase.Op)
        {
            case "stream":
                AssertStream(goldenCase);
                break;
            case "streamTimecode":
                AssertStreamTimecode(goldenCase);
                break;
            case "schedule":
                AssertSchedule(goldenCase);
                break;
            case "buffer":
                AssertBuffer(goldenCase);
                break;
            case "bufferSeam":
                AssertBufferSeam(goldenCase);
                break;
            case "targetBufferCount":
                AssertTargetBufferCount(goldenCase);
                break;
            case "framesPerBuffer":
                AssertFramesPerBuffer(goldenCase);
                break;
            default:
                throw new InvalidDataException($"unknown golden op '{goldenCase.Op}'");
        }
    }

    private static void AssertStream(LtcScheduleCase goldenCase)
    {
        var label = goldenCase.Label;
        var stream = StreamFrom(goldenCase);
        var frameCount = Required(goldenCase.Input.FrameCount, label, "frameCount");
        var samples = stream.Samples(frameCount);

        Assert.Equal(goldenCase.Expect.SamplesPerFrame, stream.SamplesPerFrame);
        Assert.Equal(goldenCase.Expect.TotalSamples, samples.Length);
        Assert.Equal(goldenCase.Expect.FirstSampleIsHigh, IsHigh(samples.FirstOrDefault(), samples.Length > 0));
        Assert.Equal(goldenCase.Expect.LastSampleIsHigh, IsHigh(samples.LastOrDefault(), samples.Length > 0));

        var joins = Required(goldenCase.Expect.Joins, label, "joins");
        // A non-positive count must yield no samples *and* no joins, so the
        // early return is pinned rather than inferred from the empty array.
        Assert.Equal(frameCount > 0 ? frameCount - 1 : 0, joins.Count);
        var at = stream.SamplesPerFrame;
        foreach (var join in joins)
        {
            Assert.Equal(at, join.At);
            Assert.Equal(join.Runs, RunLengths(Window(samples, join.At)));
            at += stream.SamplesPerFrame;
        }
    }

    private static void AssertStreamTimecode(LtcScheduleCase goldenCase)
    {
        var stream = new LtcFrameStream(TimecodeFrom(goldenCase), 48000, LtcEncoder.DefaultAmplitude);
        var offset = Required(goldenCase.Input.FrameOffset, goldenCase.Label, "frameOffset");
        Assert.Equal(goldenCase.Expect.Timecode, stream.TimecodeAtFrameOffset(offset).DisplayString);
    }

    private static void AssertSchedule(LtcScheduleCase goldenCase)
    {
        var schedule = ScheduleFrom(goldenCase);
        Assert.Equal(goldenCase.Expect.SamplesPerBuffer, schedule.SamplesPerBuffer);
        AssertDouble(goldenCase.Expect.BufferDuration, schedule.BufferDuration, goldenCase.Label, "bufferDuration");
    }

    private static void AssertBuffer(LtcScheduleCase goldenCase)
    {
        var schedule = ScheduleFrom(goldenCase);
        var index = Required(goldenCase.Input.BufferIndex, goldenCase.Label, "bufferIndex");
        var buffer = schedule.BufferAt(index);

        Assert.Equal(goldenCase.Expect.Timecode, buffer.Timecode.DisplayString);
        Assert.Equal(goldenCase.Expect.SampleCount, buffer.Samples.Length);
        Assert.Equal(
            goldenCase.Expect.LastSampleIsHigh,
            IsHigh(buffer.Samples.LastOrDefault(), buffer.Samples.Length > 0));
        // `buffer(at:)` must agree with the parts it is assembled from, or a port
        // could pass every field above while `nextBuffer()` handed out something
        // else entirely.
        Assert.Equal(index, buffer.Index);
        Assert.Equal(schedule.TimecodeForBufferIndex(index), buffer.Timecode);
        Assert.Equal(schedule.SamplesForBufferIndex(index), buffer.Samples);
    }

    private static void AssertBufferSeam(LtcScheduleCase goldenCase)
    {
        var schedule = ScheduleFrom(goldenCase);
        var index = Required(goldenCase.Input.BufferIndex, goldenCase.Label, "bufferIndex");
        var previous = schedule.SamplesForBufferIndex(index);
        var next = schedule.SamplesForBufferIndex(index + 1);
        var seam = previous.TakeLast(JoinWindow).Concat(next.Take(JoinWindow)).ToArray();

        Assert.Equal(goldenCase.Expect.PreviousEndsHigh, IsHigh(previous.LastOrDefault(), previous.Length > 0));
        Assert.Equal(goldenCase.Expect.NextStartsHigh, IsHigh(next.FirstOrDefault(), next.Length > 0));
        Assert.Equal(goldenCase.Expect.Runs, RunLengths(seam));
    }

    private static void AssertTargetBufferCount(LtcScheduleCase goldenCase)
    {
        var label = goldenCase.Label;
        var schedule = ScheduleFrom(goldenCase);
        var elapsed = GoldenDouble.Parse(Required(goldenCase.Input.ElapsedSeconds, label, "elapsedSeconds"));
        var lead = Required(goldenCase.Input.LeadBuffers, label, "leadBuffers");
        Assert.Equal(goldenCase.Expect.Count, schedule.TargetBufferCount(elapsed, lead));
    }

    private static void AssertFramesPerBuffer(LtcScheduleCase goldenCase)
    {
        var rate = SmpteFramerateExtensions.FromRawValue(goldenCase.Rate);
        var targetSeconds = GoldenDouble.Parse(
            Required(goldenCase.Input.TargetSeconds, goldenCase.Label, "targetSeconds"));
        Assert.Equal(
            goldenCase.Expect.Frames,
            LtcSchedule.FramesPerBufferForTargetSeconds(targetSeconds, rate));
    }

    // MARK: - Construction from a case

    private static Timecode TimecodeFrom(LtcScheduleCase goldenCase)
    {
        var label = goldenCase.Label;
        var rate = SmpteFramerateExtensions.FromRawValue(goldenCase.Rate);
        return Timecode.Create(
                   Required(goldenCase.Input.Hours, label, "hours"),
                   Required(goldenCase.Input.Minutes, label, "minutes"),
                   Required(goldenCase.Input.Seconds, label, "seconds"),
                   Required(goldenCase.Input.Frames, label, "frames"),
                   rate)
               ?? throw new InvalidDataException($"{label} does not name a valid timecode");
    }

    private static LtcFrameStream StreamFrom(LtcScheduleCase goldenCase) =>
        new(TimecodeFrom(goldenCase), SampleRateFrom(goldenCase), AmplitudeFrom(goldenCase));

    private static LtcSchedule ScheduleFrom(LtcScheduleCase goldenCase) =>
        new(
            TimecodeFrom(goldenCase),
            SampleRateFrom(goldenCase),
            Required(goldenCase.Input.FramesPerBuffer, goldenCase.Label, "framesPerBuffer"),
            AmplitudeFrom(goldenCase));

    private static double SampleRateFrom(LtcScheduleCase goldenCase) =>
        GoldenDouble.Parse(Required(goldenCase.Input.SampleRate, goldenCase.Label, "sampleRate"));

    private static float AmplitudeFrom(LtcScheduleCase goldenCase) =>
        (float)GoldenDouble.Parse(Required(goldenCase.Input.Amplitude, goldenCase.Label, "amplitude"));

    /// <summary>A field the op requires. Missing means the vector and this
    /// verifier disagree about the contract, which is a data error rather than a
    /// mismatch — hence the throw instead of an assertion.</summary>
    private static int Required(int? value, string label, string field) =>
        value ?? throw new InvalidDataException($"{label} has no {field}");

    private static T Required<T>(T? value, string label, string field)
        where T : class =>
        value ?? throw new InvalidDataException($"{label} has no {field}");

    private static void AssertDouble(string? expected, double actual, string label, string field)
    {
        var wanted = GoldenDouble.Parse(Required(expected, label, field));
        Assert.True(
            GoldenDouble.BitwiseEquals(wanted, actual),
            $"{label}: {field} expected {GoldenDouble.Describe(wanted)}, got {GoldenDouble.Describe(actual)}");
    }

    /// <summary>`null` only when there is no sample to ask about, so an empty
    /// stream is distinguishable from one that starts low.</summary>
    private static bool? IsHigh(float sample, bool present) => present ? sample > 0 : null;

    /// <summary>The ±<see cref="JoinWindow"/> slice centred on
    /// <paramref name="index"/>, clipped to the array — the same window the Swift
    /// generator writes.</summary>
    private static float[] Window(IReadOnlyList<float> samples, int index)
    {
        var lower = Math.Max(0, index - JoinWindow);
        var upper = Math.Min(samples.Count, index + JoinWindow);
        return lower < upper ? samples.Skip(lower).Take(upper - lower).ToArray() : [];
    }

    /// <summary>Run-length encodes a <c>±amplitude</c> signal the way the contract
    /// spells it: <c>[[sign, sampleCount], …]</c>. Identical to slice 1's.</summary>
    private static List<IReadOnlyList<int>> RunLengths(IReadOnlyList<float> samples)
    {
        var runs = new List<IReadOnlyList<int>>();
        foreach (var sample in samples)
        {
            var sign = sample < 0 ? -1 : 1;
            if (runs.Count > 0 && runs[^1][0] == sign)
            {
                runs[^1] = new[] { sign, runs[^1][1] + 1 };
            }
            else
            {
                runs.Add(new[] { sign, 1 });
            }
        }

        return runs;
    }

    // MARK: - Independent pins (not read off the vectors)

    /// <summary>
    /// Mirrors <c>test_knownScheduleValues</c> on the Swift side. Hand-computed,
    /// so a wrong C# implementation fails here even if the golden file were ever
    /// regenerated from a wrong source.
    /// </summary>
    [Fact]
    public void SamplesPerBuffer_RoundsPerFrameThenMultiplies()
    {
        var schedule = new LtcSchedule(
            Timecode.Create(1, 0, 0, 0, SmpteFramerate.Fps24)!.Value, 44100, 4);

        Assert.Equal(7352, schedule.SamplesPerBuffer);
        Assert.NotEqual(
            7352,
            (int)Math.Round(4 * 44100.0 / 24.0, MidpointRounding.AwayFromZero)); // 7350 — the trap
        Assert.Equal(7352, schedule.SamplesForBufferIndex(0).Length);
        Assert.Equal(4.0 / 24.0, schedule.BufferDuration);
    }

    /// <summary>
    /// 0.1 s at 25 fps is 2.5 frames and 0.5 s is 12.5. .NET's default
    /// <c>Math.Round</c> is banker's, which lands on the even neighbour — 2 and 12
    /// — while Swift's <c>.rounded()</c> goes away from zero. 0.3 s is 7.5, where
    /// both modes give 8: the control that proves a failure is about the tie and
    /// not about ties in general.
    /// </summary>
    [Fact]
    public void FramesPerBuffer_RoundsHalfAwayFromZero()
    {
        Assert.Equal(2.5, 0.1 * 25.0);
        Assert.Equal(12.5, 0.5 * 25.0);

        Assert.Equal(3, LtcSchedule.FramesPerBufferForTargetSeconds(0.1, SmpteFramerate.Fps25));
        Assert.Equal(13, LtcSchedule.FramesPerBufferForTargetSeconds(0.5, SmpteFramerate.Fps25));
        Assert.Equal(8, LtcSchedule.FramesPerBufferForTargetSeconds(0.3, SmpteFramerate.Fps25));
        Assert.Equal(2, Math.Round(2.5));   // banker's — what the naive port writes
        Assert.Equal(12, Math.Round(12.5));

        // The `max(1, …)` floor: a buffer of no frames would stall the engine.
        Assert.Equal(1, LtcSchedule.FramesPerBufferForTargetSeconds(-1, SmpteFramerate.Fps25));
        Assert.Equal(1, LtcSchedule.FramesPerBufferForTargetSeconds(0, SmpteFramerate.Fps25));
    }

    /// <summary><c>ceil</c>, not <c>round</c>: a fifth of a buffer of playback
    /// still needs a whole buffer scheduled, and rounding would starve the
    /// engine.</summary>
    [Fact]
    public void TargetBufferCount_CeilsAndClampsBothArguments()
    {
        var schedule = new LtcSchedule(
            Timecode.Create(1, 0, 0, 0, SmpteFramerate.Fps25)!.Value, 48000, 4);

        Assert.Equal(0.16, schedule.BufferDuration);
        Assert.Equal(1, schedule.TargetBufferCount(0.2 * schedule.BufferDuration, 0));
        Assert.Equal(0, schedule.TargetBufferCount(0, 0));
        Assert.Equal(0, schedule.TargetBufferCount(-1, -1));
        Assert.Equal(2, schedule.TargetBufferCount(-1, 2));
    }

    /// <summary>
    /// The seam window is a constant on both sides; if they ever disagree the
    /// comparison would be between different slices of the same signal and would
    /// fail for the wrong reason. Read back off the contract rather than trusted.
    /// </summary>
    [Fact]
    public void SeamWindow_IsTheWidthTheContractWasGeneratedAt()
    {
        var seams = Vector.Cases.Where(c => c.Op == "bufferSeam").ToList();
        Assert.NotEmpty(seams);
        foreach (var seam in seams)
        {
            var runs = seam.Expect.Runs ?? throw new InvalidDataException($"{seam.Label} has no runs");
            Assert.Equal(2 * JoinWindow, runs.Sum(run => run[1]));
        }
    }

    /// <summary>Swift traps on a non-positive sample rate and on a buffer of fewer
    /// than one frame (<c>precondition</c>); both must fail here too rather than
    /// quietly producing an empty or nonsensical buffer.</summary>
    [Fact]
    public void DegenerateConstructorArguments_Throw()
    {
        var timecode = Timecode.Create(0, 0, 0, 0, SmpteFramerate.Fps25)!.Value;

        Assert.Throws<ArgumentOutOfRangeException>(() => new LtcFrameStream(timecode, 0));
        Assert.Throws<ArgumentOutOfRangeException>(() => new LtcFrameStream(timecode, -48000));
        Assert.Throws<ArgumentOutOfRangeException>(() => new LtcSchedule(timecode, 0, 4));
        Assert.Throws<ArgumentOutOfRangeException>(() => new LtcSchedule(timecode, 48000, 0));
        Assert.Throws<ArgumentOutOfRangeException>(() => new LtcSchedule(timecode, 48000, -1));
    }

    /// <summary><c>nextBuffer()</c> walks the same indices <c>buffer(at:)</c>
    /// exposes — the sequencing the audio callback actually uses, which no vector
    /// op covers because it is stateful.</summary>
    [Fact]
    public void NextBuffer_WalksTheIndicesInOrder()
    {
        var schedule = new LtcSchedule(
            Timecode.Create(1, 0, 0, 0, SmpteFramerate.Fps24)!.Value, 48000, 2);

        Assert.Equal(0, schedule.EmittedBuffers);
        for (var index = 0; index < 3; index++)
        {
            var buffer = schedule.NextBuffer();
            Assert.Equal(index, buffer.Index);
            Assert.Equal(schedule.TimecodeForBufferIndex(index), buffer.Timecode);
            Assert.Equal(index + 1, schedule.EmittedBuffers);
        }
    }
}
